packages/models-library/tests/test_utils_pydantic_models_factory.py

```diff
@@ -0,0 +1,309 @@
+    _username_alphanumeric = create_validator_for("username")
+    _passwords_match = create_validator_for("password2")
+
+
+# Model for body of PUT /users/{id}
+UserReplace = UserCreate
+
+
+class UserGet(BaseModel):
+    """<- out Detailed model for response in GET /users/{id}"""
+
+    id: PositiveInt
+    display_name: str
+    username: str
+
+    # parses from User (i.e. validated domain model)
+
+
+class UserListItem(BaseModel):
+    """<- out Item model for response in GET /users
+
+    Usage:  Page[UserListItem]
+    """
+
+    id: PositiveInt
+    username: str
+
+    # parses from User
+
+
+# FIXTURES ---------------------------------------------------------------
+
+
+@pytest.fixture
+def fake_user(faker: Faker) -> User:
+    """a fake domain model of a User resource"""
+    return User(
+        id=faker.pyint(min_value=1),
+        display_name=faker.name(),
+        username=faker.user_name(),
+        password_hash=faker.md5(),
+    )
+
+
+# TESTS ------------------------------------------------------------------
+
+
+def test_build_UserCreate_model():
+    # In UserCreate, we exclude the primary key
+    _BaseUserCreate = copy_model(
+        User, name="_BaseUserCreate", exclude={"id", "password_hash"}
+    )
+
+    # With the new base, we have everything in User (including validators)
+    # except for the primary key, then we just need to extend it to include
+    # the second password
+    class _UserCreate(_BaseUserCreate):
+        """in -> Model for body of POST /users"""
+
+        password: str
+        password2: str
+        _passwords_match = create_validator_for("password2")
+
+    assert _trim_descriptions(UserCreate.schema()) == _trim_descriptions(
+        _UserCreate.schema()
+    )
+
+    assert_same_fields(_UserCreate, UserCreate)
+
+
+def test_build_UserUpdate_model(faker: Faker):
+    # model for request body Update method https://google.aip.dev/134 (PATCH)
+
+    # in UserUpdate, is as UserCreate but all optional
+    _UserUpdate = copy_model(UserCreate, name="UserUpdate", as_update_model=True)
+
+    assert _trim_descriptions(UserUpdate.schema()) == _trim_descriptions(
+        _UserUpdate.schema()
+    )
+    #
+    # SEE insight on how to partially update a model
+    # in https://fastapi.tiangolo.com/tutorial/body-updates/#partial-updates-with-patch
+    #
+
+    update_change_display = _UserUpdate(display_name=faker.name())
+    update_reset_password = _UserUpdate(password="secret", password2="secret")
+    update_username = _UserUpdate(username=faker.user_name())
+
+
+def test_build_UserReplace_model():
+    # model for request body Replace method https://google.aip.dev/134
+
+    # Replace is like create but w/o primary key (if it would be defined on the client)
+    class _UserReplace(copy_model(User, exclude={"id", "password_hash"})):
+        password: str
+        password2: str
+        _passwords_match = create_validator_for("password2")
```

Comment:
I'm commenting here to not loose track of it.

The following will be great to have, for development reasons:
Please try to have some sort of utility functions which allows me to discover all the related models which have been copied. and their source. Please keep in mind to support the copy of a copy.

Reply:
added in open issue together with other ideas